README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# moostr <a href="https://iris-yi-jiang.github.io/moostr/"><img src="man/figures/logo.png" align="right" height="139" /></a>

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

## Installation

You can install the development version of moostr via:

``` r
# install.packages("remotes")
remotes::install_github("iris-yi-jiang/moostr")
```

## Overview

`moostr` generates a Moodle syntax string for a given list of possible
answers and their corresponding rewards. It allows the user to
conveniently allocate rewards for multiple correct and partially correct
answers. For multichoice questions, `moostr` also randomly shuffles the
answer list. For numerical questions, `moostr` also allows the
flexibility to specify different tolerance levels for each possible
answer.

For `moostr` to function properly, it is necessary to set the question
type for both multichoice and numerical questions to vertatim type. The
returned Moodle syntax string can then be directly used as the solution
to a verbatim item in an Embedded Answers (Cloze) exercise with `exams`
or Moodle.

## Example

This is a basic example for mutichoice question type. Note that `moostr`
shuffles the answer list.

``` r
library(moostr)
make_moostr(type="mchoice", 
            ans=c("A two-sample t-test",
                  "A paired t-test", 
                  "A one-sample t-test"), 
            reward=c(50, 100, 0))
#> [1] ":MULTICHOICE:%100%A paired t-test~%50%A two-sample t-test~%0%A one-sample t-test"
```

This is a basic example for numerical question type. Note that `moostr`
allows the flexibly to specify different tolerance levels for each
possible answer. If the same tolerance level is preferred across all
possible answers, only a single tolerance level is needed for the tol
argument. The tol argument is by default set to 0.01, so if this is the
preferred tolerance level, then it is not necessary to set it.

The tol argument should be set carefully. If the difference between
correct and partially correct answers are within the specified tolerance
level, then the reward applied to the partially correct answer will be
ignored and the reward for the correct answer will be used to evaluate
both answers.

``` r
library(moostr)
make_moostr(type="num", 
            ans=c(2, 2.1, 2.01), 
            reward=c(30, 100, 50),
            tol = c(0, 0.1, 0.01))
#> [1] ":NUMERICAL:%30%2:0~%100%2.1:0.1~%50%2.01:0.01"
make_moostr(type="num", 
            ans=c(2, 2.1, 2.01), 
            reward=c(30, 100, 50),
            tol = 0.1)
#> [1] ":NUMERICAL:%30%2:0.1~%100%2.1:0.1~%50%2.01:0.1"
make_moostr(type="num", 
            ans=c(2, 2.1, 2.01), 
            reward=c(30, 100, 50))
#> [1] ":NUMERICAL:%30%2:0.01~%100%2.1:0.01~%50%2.01:0.01"
```


NEWS.md
# moostr 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(moostr)

test_check("moostr")


tests/testthat/test-make_moostr.R
test_that("test output of make_moostr", {
    set.seed(2022)
    expect_identical(make_moostr(type="mchoice",
                                 ans=c("a", "b", "c"),
                                 reward=c(50, 100, 0)),
                     ":MULTICHOICE:%0%c~%100%b~%50%a")
    set.seed(2022)
    expect_identical(make_moostr(type="num",
                                 ans=c(2, 3, 4.5),
                                 reward=c(0, 100, 50)),
                     ":NUMERICAL:%0%2:0.01~%100%3:0.01~%50%4.5:0.01")
    set.seed(2022)
    expect_error(make_moostr(type="num",
                             ans= c(1),
                             reward= c(100)),
                 paste("make_moostr is meant for multiple answers",
                       "so length\\(ans\\) should be greater than one"))
    set.seed(2022)
    expect_error(make_moostr(type="num",
                             ans= c(2, 3, 4.5),
                             reward= c(0, 200, 50)),
                 "at least one reward has to be 100")
    set.seed(2022)
    expect_error(make_moostr(type="shortanswer",
                 ans=c(2, 3, 4.5),
                 reward=c(0, 100, 50)),
                 "type can only be 'mchoice' or 'num'")
    set.seed(2022)
    expect_error(make_moostr(type="num",
                             ans=c(2, 3, 4.5),
                             reward=c(0, 100, 50, 30)),
                 "lengh of ans and reward must match")
})


R/make_moostr.R
#' Moodle Syntax String for Multichoice and Numerical Question Types
#'
#' @description This package generates a Moodle syntax string for
#'   a given list of possible answers and their corresponding rewards.
#'   It allows the user to conveniently allocate rewards for multiple
#'   correct and partially correct answers For multichoice questions, the
#'   package also randomly shuffles the answer list. For numerical questions,
#'   the package also allows the flexibility to specify different tolerance
#'   levels for each possible answer.
#'
#'   For the package to function properly, it is necessary to set the
#'   question type for both multichoice and numerical questions to vertatim
#'   type.The returned Moodle syntax string can then be directly used as the
#'   solution to a verbatim item in an Embedded Answers (Cloze) exercise
#'   with exams package or Moodle.
#'
#' @param type string; either "mchoice" or "num"
#' @param ans a list of possible answers; for "mchoice" type, each
#'   possible answer is a string; for "num" type, each possible answer is
#'   of numeric type
#' @param reward a list of corresponding rewards for each possible
#'   answer expressed in percentages (0 and 100 inclusive)
#' @param tol a tolerance level for "num" type; this argument is
#'   ignored for "mchoice" type
#' @return a Moodle syntax string
#' @export
#'
#' @examples
#' #For multichoice type questions
#'
#' make_moostr(type="mchoice",
#'             ans=c("A two-sample t-test",
#'                   "A paired t-test",
#'                   "A one-sample t-test"),
#'             reward=c(50, 100, 0))
#'
#' #For numerical type questions
#'
#' make_moostr(type="num",
#'             ans=c(2, 2.1, 2.01),
#'             reward=c(30, 100, 50),
#'             tol = c(0, 0.1, 0.01))
#'
#' make_moostr(type="num",
#'             ans=c(2, 2.1, 2.01),
#'             reward=c(50, 100, 50),
#'             tol = 0)
#'
#' make_moostr(type="num",
#'             ans=c(2, 2.1, 2.01),
#'             reward=c(30, 100, 50))

make_moostr <- function(type, ans, reward, tol = 0.01) {
    if (length(ans) <= 1){
        stop(paste("make_moostr is meant for multiple answers",
                   "so length(ans) should be greater than one"))
    }
    if (max(reward) != 100){
        stop("at least one reward has to be 100")
    }
    if (!(type %in% c("mchoice", "num"))){
        stop("type can only be 'mchoice' or 'num'")
    }
    if (length(ans) != length(reward)){
        stop("lengh of ans and reward must match")
    }
    if (type == "mchoice"){
        s <- make_moostr_multichoice(ans, reward)
    } else {
        s <- make_moostr_numerical(ans, reward, tol = tol)
    }
    return(s)
}

make_moostr_multichoice <- function(ans, reward){
    idx <- sample(seq_along(ans))
    ans <- ans[idx]
    reward <- reward[idx]
    temp <- glue::glue("%{reward}%{ans}")
    multi_reward <- glue::glue_collapse(temp, sep = "~", last= "")
    s <- paste0(":MULTICHOICE:", multi_reward)
    return(s)
}

make_moostr_numerical <- function(ans, reward, tol = 0.01){
    temp <- glue::glue("%{reward}%{ans}:{tol}")
    multi_reward <- glue::glue_collapse(temp, sep = "~", last= "")
    s <- paste0(":NUMERICAL:", multi_reward)
    return(s)
}
